File: server/src/main/java/com/chenkh/vchat/server/msgtaskaccess/MsgMgr.java
package com.chenkh.vchat.server.msgtaskaccess;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.nio.ByteBuffer;
import java.nio.channels.AsynchronousSocketChannel;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


import com.chenkh.vchat.base.msg.*;
import com.chenkh.vchat.base.msg.both.DeleteFriendMsg;
import com.chenkh.vchat.base.msg.both.UserStateChangeMsg;
import com.chenkh.vchat.base.msg.client.AddFriendMsg;
import com.chenkh.vchat.base.msg.client.LoginMsg;
import com.chenkh.vchat.base.msg.client.QueryMsg;
import com.chenkh.vchat.base.msg.client.RegisterMsg;
import com.chenkh.vchat.base.msg.server.AddFriendSucess;
import com.chenkh.vchat.base.msg.server.LoginResultMsg;
import com.chenkh.vchat.base.msg.server.QueryResultMsg;
import com.chenkh.vchat.base.msg.server.RegisterSucessMsg;
import com.chenkh.vchat.base.msg.server.StrangerMsg;
import com.chenkh.vchat.base.msg.server.enu.LoginResult_Type;
import com.chenkh.vchat.server.dao.GroupDAO;
import com.chenkh.vchat.server.dao.MsgDAO;
import com.chenkh.vchat.server.dao.Mysql_GroupDAO;
import com.chenkh.vchat.server.dao.Mysql_MsgDAO;
import com.chenkh.vchat.server.exception.PasswordNotCorrect;
import com.chenkh.vchat.server.exception.RegiserUserFailedException;
import com.chenkh.vchat.server.exception.UserNotFoundException;
import com.chenkh.vchat.server.msgtaskaccess.client.Client;
import com.chenkh.vchat.base.bean.Friend;
import com.chenkh.vchat.base.bean.Group;
import com.chenkh.vchat.base.bean.Stranger;
import com.chenkh.vchat.base.bean.User;
import com.chenkh.vchat.base.bean.VState;

public class MsgMgr implements MsgClientAccess {
	private final UserAccess userMgr;
	private ClientAccess clientMgr;
	private TaskAccessMgr taskMgr;
	private MsgDAO msgDao = new Mysql_MsgDAO();
	private GroupDAO groupDao = new Mysql_GroupDAO();

	public MsgMgr(TaskAccessMgr mgr) {
		userMgr = new UserMgr();
		this.taskMgr = mgr;
		clientMgr = new ClientMgr();
	}

	@Override
	public void parseMsg(ClientMsg msg, AsynchronousSocketChannel socket) {
		msg.parse(this, socket);

	}



	@Override
	public void parseMsg(LoginMsg msg, AsynchronousSocketChannel socket) {
		int id = msg.getId();
		String password = msg.getPassword();

		ServerMsg serverMsg = null;
		List<ContenMsg> offlineMsgs = null;
		ServerMsg offlineMsg = null;

		try {
			User u = null;
			u = userMgr.getUser(id, password);
			serverMsg = new LoginResultMsg(LoginResult_Type.SUCCESS, u);
			Set<Integer> friendIds = new HashSet<Integer>();
			for (Group group : u.getGroups()) {
				for (Friend friend : group.getFriends()) {
					if (clientMgr.isContain(friend.getId())) {
						VState mstate = clientMgr.getState(friend.getId());
						friend.setState(mstate);
					}
					friendIds.add(friend.getId());
				}
			}

			VState state = msg.getState();
			System.out.println("state:" + state);
			clientMgr.addClient(id, socket, state, friendIds);

			this.parseMsg(new UserStateChangeMsg(id, state), socket);

			offlineMsgs = msgDao.getOfflineMsg(u.getId());
			if (offlineMsgs != null && offlineMsgs.size() > 0) {

				offlineMsg = new OfflineMsg(u.getId(), offlineMsgs);
			}

		} catch (UserNotFoundException e) {
			serverMsg = new LoginResultMsg(LoginResult_Type.USER_NOTFOUN, null);

		} catch (PasswordNotCorrect e) {
			serverMsg = new LoginResultMsg(
					LoginResult_Type.PASSWORD_NOTCORRECT, null);

		} finally {
			ByteBuffer bf = this.parseObject(serverMsg);
			taskMgr.addWriteTask(bf, socket);
			if (offlineMsg != null) {
				ByteBuffer bf2 = this.parseObject(offlineMsg);
				taskMgr.addWriteTask(bf2, socket);
			}

		}

	}





	@Override
	public void parseMsg(ChatMsg msg, AsynchronousSocketChannel socket) {

		int fromId = msg.getFromId();
		int toId = msg.getToId();
		// 如果存在，则发送消息，否则，转存为离线消息
		if (clientMgr.isContain(toId)) {
			Client client = clientMgr.getClient(toId);
			AsynchronousSocketChannel so = client.getSocket();
			if (clientMgr.isFriend(toId, fromId)) {

				ByteBuffer bf = this.parseObject(msg);
				if (bf != null) {
					taskMgr.addWriteTask(bf, so);
				}
			} else {
				Stranger stranger = userMgr.getStrangerById(fromId);
				String conten = msg.getContent();
				ServerMsg smsg = new StrangerMsg(stranger, conten, toId);
				ByteBuffer bf = this.parseObject(smsg);
				taskMgr.addWriteTask(bf, so);
			}

		} else {
			msgDao.addOfflineMsg(fromId, toId, msg.getContent(), msg.getDate());
		}

	}

	// 传入对象，转换成bytebuffer
	private ByteBuffer parseObject(Object msg) {

		ByteArrayOutputStream bos = new ByteArrayOutputStream();
		ByteBuffer bf = null;

		try (ObjectOutputStream oos = new ObjectOutputStream(bos)) {
			oos.writeObject(msg);
			byte[] bytes = bos.toByteArray();
			bf = ByteBuffer.wrap(bytes);

		} catch (IOException e) {
			System.out.println("转换失败！");
			e.printStackTrace();
		}

		return bf;
	}

	@Override
	public void parseMsg(VerifyMsg msg, AsynchronousSocketChannel socket) {
		// TODO Auto-generated method stub
		System.out.println("test");

	}

	@Override
	public void parseMsg(UserStateChangeMsg msg,
			AsynchronousSocketChannel socket) {
		int id = msg.getFromId();
		if (clientMgr.isContain(id)) {
			Client c = clientMgr.getClient(id);
			c.setState(msg.getState());
		}

		Set<Integer> friends = clientMgr.getFriends(id);
		for (Integer friendid : friends) {
			if (clientMgr.isContain(friendid)) {
				Client c = clientMgr.getClient(friendid);
				ByteBuffer bf = this.parseObject(msg);
				taskMgr.addWriteTask(bf, c.getSocket());
			}
		}

	}

	@Override
	public void reMoveClient(AsynchronousSocketChannel socket) {
		int id = clientMgr.removeCient(socket);
		if (id != -1) {
			this.parseMsg(new UserStateChangeMsg(id, VState.offline), socket);
		}

	}

	@Override
	public void parseMsg(RegisterMsg msg, AsynchronousSocketChannel socket) {
		User u = msg.getUser();
		int id = -1;
		try {
			id = userMgr.addUser(u);
			if (id != -1) {
				ServerMsg resultMsg = new RegisterSucessMsg(id);
				ByteBuffer bf = this.parseObject(resultMsg);
				taskMgr.addWriteTask(bf, socket);
			}
		} catch (RegiserUserFailedException e) {

		}

	}

	@Override
	public void parseMsg(QueryMsg msg, AsynchronousSocketChannel socket) {
		int toId = msg.getQueryId();
		String keyword = msg.getKeyword();
		List<Stranger> strangers = userMgr.queryUser(toId, keyword);
		ServerMsg smsg = new QueryResultMsg(toId, strangers);
		ByteBuffer bf = this.parseObject(smsg);
		taskMgr.addWriteTask(bf, socket);
	}

	@Override
	public void parseMsg(AddFriendMsg msg, AsynchronousSocketChannel socket) {
		int fromId = msg.getMyId();
		int friendId = msg.getFriendId();
		String noteName = msg.getNoteName();
		int groupId = msg.getGroupId();
		Friend friend = userMgr.addFriend(fromId, friendId, groupId, noteName);

		if (friend != null) {
			if (clientMgr.isContain(friendId)) {
				Client friendClient = clientMgr.getClient(friendId);
				friend.setState(friendClient.getState());
			}

			if (clientMgr.isContain(fromId)) {
				clientMgr.addFriend(fromId, friendId);
				ServerMsg suceMsg = new AddFriendSucess(friend, groupId);
				ByteBuffer bf = this.parseObject(suceMsg);
				taskMgr.addWriteTask(bf, socket);
			}

		}

	}

	@Override
	public void parseMsg(DeleteFriendMsg msg, AsynchronousSocketChannel socket) {
		int myId = msg.getMyId();
		int friendId = msg.getFriendId();
		int groupId = msg.getGroupId();
		boolean result = userMgr.deleteFriend(myId, friendId, groupId);

		if (result) {
			ByteBuffer bf = this.parseObject(msg);
			taskMgr.addWriteTask(bf, socket);

		}

	}

}

File: server/src/main/java/com/chenkh/vchat/server/manager/MsgMgr.kt
package com.chenkh.vchat.server.manager

import com.chenkh.vchat.base.bean.MsgType.Client2Server
import com.chenkh.vchat.base.bean.User
import com.chenkh.vchat.base.bean.VState
import com.chenkh.vchat.base.msg.*
import com.chenkh.vchat.base.msg.both.DeleteFriendMsg
import com.chenkh.vchat.base.msg.both.UserStateChangeMsg
import com.chenkh.vchat.base.msg.client.AddFriendMsg
import com.chenkh.vchat.base.msg.client.LoginMsg
import com.chenkh.vchat.base.msg.client.QueryMsg
import com.chenkh.vchat.base.msg.client.RegisterMsg
import com.chenkh.vchat.base.msg.server.*
import com.chenkh.vchat.server.dao.GroupDAO
import com.chenkh.vchat.server.dao.MsgDAO
import com.chenkh.vchat.server.dao.Mysql_GroupDAO
import com.chenkh.vchat.server.dao.Mysql_MsgDAO
import com.chenkh.vchat.server.exception.PasswordNotCorrect
import com.chenkh.vchat.server.exception.RegiserUserFailedException
import com.chenkh.vchat.server.exception.UserNotFoundException
import com.chenkh.vchat.server.net.IChannel
import com.chenkh.vchat.server.net.IChannelConnectListener
import com.chenkh.vchat.server.net.IMsgListener
import com.chenkh.vchat.server.net.IServer
import java.nio.channels.AsynchronousSocketChannel
import java.util.*


class MsgMgr(server: IServer) : IMsgListener, IChannelConnectListener {
    private val userMgr: UserAccess
    private val clientMgr: IClientManager

    //private TaskAccessMgr taskMgr;
    private val msgDao: MsgDAO = Mysql_MsgDAO()
    private val groupDao: GroupDAO = Mysql_GroupDAO()

    //@Override
    fun parseLoginMsg(msg: LoginMsg, channel: IChannel?) {
        val id = msg.id
        val password = msg.password

        //ServerMsg<LoginResultMsg> serverMsg=null;
        var loginResultMsg: LoginResultMsg? = null
        var offlineMsgs: List<ContenMsg?>? = null
        var offlineMsg: OfflineMsg? = null
        try {
            var u: User? = null
            u = userMgr.getUser(id, password)

            //serverMsg = new LoginResultMsg(LoginResult_Type.SUCCESS, u);
            val friendIds: MutableSet<Int> = HashSet()
            for (group in u!!.groups) {
                for (friend in group.friends) {
                    if (clientMgr.isContain(friend.id)) {
                        val mstate = clientMgr.getState(friend.id)
                        friend.state = mstate
                    }
                    friendIds.add(friend.id)
                }
            }
            val state = msg.state
            println("state:$state")
            clientMgr.addClient(id, channel, state, friendIds)
            this.parseMsg(UserStateChangeMsg(id, state), channel)
            offlineMsgs = msgDao.getOfflineMsg(u.id)
            if (offlineMsgs != null && offlineMsgs.size > 0) {
                offlineMsg = OfflineMsg(u.id, offlineMsgs)
                //offlineMsg = Msg.buildServerMsg(MsgType.Server2Client.OFFLINE_MSG,);
            }
            loginResultMsg = LoginResultMsg(null, u, true)
        } catch (e: UserNotFoundException) {
            loginResultMsg = LoginResultMsg("用户未找到", null, false)
            //serverMsg = Msg.buildServerMsg(MsgType.Server2Client.LOGIN,false,"用户未找到",null);
        } catch (e: PasswordNotCorrect) {
            loginResultMsg = LoginResultMsg("密码不正确", null, false)
            //serverMsg = Msg.buildServerMsg(MsgType.Server2Client.LOGIN,false,"密码不正确",null);
        } finally {
            channel!!.sendLoginResultMsg(loginResultMsg)
            if (offlineMsg != null) {
                channel.sendOfflineMsg(offlineMsg)
            }
        }
    }

    fun parseMsg(msg: ChatMsg, channel: IChannel?) {
        val fromId = msg.fromId
        val toId = msg.toId
        // 如果存在，则发送消息，否则，转存为离线消息
        if (clientMgr.isContain(toId)) {
            val client = clientMgr.getClient(toId)
            if (clientMgr.isFriend(toId, fromId)) {
                channel!!.sendChatMsg(msg)
            } else {
                val stranger = userMgr.getStrangerById(fromId)
                val conten = msg.content
                val smsg = StrangerMsg(stranger, conten, toId)
                channel!!.sendStrangerMsg(smsg)
            }
        } else {
            msgDao.addOfflineMsg(fromId, toId, msg.content, msg.date)
        }
    }

    /*	private ByteBuffer object2ByteArrayByObjSeria(Object msg) {

		ByteArrayOutputStream bos = new ByteArrayOutputStream();
		ByteBuffer bf = null;

		try (ObjectOutputStream oos = new ObjectOutputStream(bos)) {
			oos.writeObject(msg);
			byte[] bytes = bos.toByteArray();
			bf = ByteBuffer.wrap(bytes);

		} catch (IOException e) {
			System.out.println("转换失败！");
			e.printStackTrace();
		}

		return bf;
	}*/
    fun parseMsg(msg: VerifyMsg?, socket: AsynchronousSocketChannel?) {
        // TODO Auto-generated method stub
        println("test")
    }

    fun parseMsg(msg: UserStateChangeMsg,
                 socket: IChannel?) {
        val id = msg.fromId
        if (clientMgr.isContain(id)) {
            val c = clientMgr.getClient(id)
            c!!.state = msg.state
        }
        val friends = clientMgr.getFriends(id)
        for (friendid in friends!!) {
            if (clientMgr.isContain(friendid!!)) {
                val c = clientMgr.getClient(friendid)
                c!!.socket!!.sendFriendStateChanged(msg)
            }
        }
    }

    /*	@Override
	public void reMoveClient(IChannel socket) {
		int id = clientMgr.removeCient(socket);
		if (id != -1) {
			this.parseMsg(new UserStateChangeMsg(id, VState.offline), socket);
		}

	}*/
    fun parseMsg(msg: RegisterMsg, channel: IChannel?) {
        val u = msg.user
        var id = -1
        try {
            id = userMgr.addUser(u)
            if (id != -1) {
                val resultMsg = RegisterSucessMsg(id)
                channel!!.sendRegisterSuccessMsg(resultMsg)
            }
        } catch (e: RegiserUserFailedException) {
        }
    }

    fun parseMsg(msg: QueryMsg, channel: IChannel?) {
        val toId = msg.queryId
        val keyword = msg.keyword
        val strangers = userMgr.queryUser(toId, keyword)
        val smsg = QueryResultMsg(toId, strangers)
        channel!!.sendQueryResultMsg(smsg)
    }

    fun parseMsg(msg: AddFriendMsg, channel: IChannel?) {
        val fromId = msg.myId
        val friendId = msg.friendId
        val noteName = msg.noteName
        val groupId = msg.groupId
        val friend = userMgr.addFriend(fromId, friendId, groupId, noteName)
        if (friend != null) {
            if (clientMgr.isContain(friendId)) {
                val friendClient = clientMgr.getClient(friendId)
                friend.state = friendClient!!.state
            }
            if (clientMgr.isContain(fromId)) {
                clientMgr.addFriend(fromId, friendId)
                val suceMsg = AddFriendSucess(friend, groupId)
                channel!!.sendAddFriendResultMsg(suceMsg)
            }
        }
    }

    fun parseMsg(msg: DeleteFriendMsg, channel: IChannel?) {
        val myId = msg.myId
        val friendId = msg.friendId
        val groupId = msg.groupId
        val result = userMgr.deleteFriend(myId, friendId, groupId)
        if (result) {
            channel!!.sendDeleteFriendResultMsg(msg)
        }
    }

    override fun onReceivedMsg(channel: IChannel, serverMsg: ClientMsg<*>) {
        val body = serverMsg.body
        val msgType = serverMsg.msgType
        when (msgType) {
            Client2Server.LOGIN -> parseLoginMsg(body as LoginMsg, channel)
            Client2Server.QUERY -> parseMsg(body as QueryMsg, channel)
            Client2Server.REGISTER -> parseMsg(body as RegisterMsg, channel)
            Client2Server.ADD_FRIEND -> parseMsg(body as AddFriendMsg, channel)
            Client2Server.DELETE_FRIEND -> parseMsg(body as DeleteFriendMsg, channel)
            Client2Server.USER_STATE_CHANGED -> {
                parseMsg(body as UserStateChangeMsg, channel)
                parseMsg(body as ChatMsg, channel)
            }
            Client2Server.CHAT -> parseMsg(body as ChatMsg, channel)
        }
    }

    override fun channelConnected(channel: IChannel) {}
    override fun channelDisconnect(channel: IChannel, exc: Throwable?) {
        val id = clientMgr.removeCient(channel)
        if (id != -1) {
            this.parseMsg(UserStateChangeMsg(id, VState.offline), channel)
        }
    }

    init {
        userMgr = UserMgr()
        clientMgr = ClientMgr()
     /*   server.addMsgListener(IMsgListener { channel: IChannel?, serverMsg: ClientMsg<*> -> this.onReceivedMsg(channel, serverMsg) })
        server.addMsgListener(object : IMsgListener {
            override fun onReceivedMsg(channel: IChannel, serverMsg: ClientMsg<*>) {

            }
        });
        */
        server.addMsgListener(object : IMsgListener{
            override fun onReceivedMsg(channel: IChannel, serverMsg: ClientMsg<*>) {
                this.onReceivedMsg(channel, serverMsg)
            }
        })
        server.addConnectListener(this)
    }


}